### PR TITLE
Clean code

### PR DESCRIPTION
**Pull request by SaraMahmoodabadi**
cleaned up the phase 1 code and used interfaces and manager

### CODE REVIEW

**Comment by Ali-Athary** [comment]
nice jobe! everything was ok. 👌

**Comment by Hossein-29** [comment]
1. ListCategory class isn't applying second principle of SOLID cause maybe we have other filters and then you should modify these methods to add that filter.

2.As I said before It's better to have an abstraction for docReader cause maybe we have another kind of doc or maybe we want to read it in another way so here you aren't applying fifth principle of SOLID your code should depend of on abstractions not concretes.

File: .gitignore.txt
Comment (Amirack2003) [inline .gitignore.txt:1]: please ignore uploading unnecessary files and data.(ignore book test files) 
.txt??? :/

File: src/searchMode/Search.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import dataStructures.InvertedIndex;
+import java.util.Set;
+
+public abstract class Search {
+    protected final String query;
+    protected InvertedIndex database;
+
+    public Search(InvertedIndex database, String query) {
+        this.query = query;
+        this.database = database;
```
Comment (Amirack2003) [inline src/searchMode/Search.java]: database? you are not using any information from any database please change the name

File: src/searchMode/advancedSearch/AdvancedSearch.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class AdvancedSearch extends Search {
+    private final Set<String> queryWords;
+    private final ListCategory listCategory;
+
+
+    public AdvancedSearch(InvertedIndex database, String query) {
+        super(database, query);
```
Comment (Amirack2003) [inline src/searchMode/advancedSearch/AdvancedSearch.java]: .

File: src/searchMode/advancedSearch/AdvancedSearch.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        queryWords = new HashSet<>();
+        listCategory = new ListCategory();
+    }
+
+
+    public void addToListCategory(String word, ListType type) {
```
Comment (Amirack2003) [inline src/searchMode/advancedSearch/AdvancedSearch.java:27]: what if we would also have query start with ^query indicates that we prefer documents with query over others? this structure does not handle this and it is in contradiction with SOLID (O)

File: textFiles/The Clean Coder.txt [hunk 4]
```diff
@@ -0,0 +1,157 @@
+The Clean Coder Summary
```
Comment (Amirack2003) [inline textFiles/The Clean Coder.txt:1]: git ignore T_T

File: src/searchMode/NormalSearch.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import dataStructures.InvertedIndex;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class NormalSearch extends Search {
+    private Set<String> queryWords;
+
+    public NormalSearch(InvertedIndex database, String query) {
```
Comment (Amirack2003) [inline src/searchMode/NormalSearch.java]: database?

File: src/reader/IReader.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package reader;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class IReader {
```
Comment (Amirack2003) [inline src/reader/IReader.java]: I indicates for interface and it is name convention in C# not Java. Even if then you should use AbstractReader but i do not recommend that.

File: src/reader/DocReader.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        super();
+        this.filesPath = filesPath;
+    }
+
+    private File[] getFiles() {
+        File directoryPath = new File(filesPath);
+        return directoryPath.listFiles();
+    }
+
+    private StringBuilder getFileLines(File file) {
```
Comment (Amirack2003) [inline src/reader/DocReader.java]: There is a class with name StringBuilder in java, please rename this.(it also should be in lamelCase format).
In some of your code function name are in lamelCase and other function name is not. Please follow a specific name format

File: src/FileManager.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import filter.normalizer.UpperCaseNormalizer;
+import filter.tokenizer.SplitTokenizer;
+import filter.tokenizer.Tokenizer;
+import reader.IReader;
+
+import java.util.Map;
+
+public class FileManager {
+    private final InvertedIndex invertedIndex;
+    private final Normalizer normalizer;
```
Comment (MSPoulaei) [inline src/FileManager.java]: - FileManager class should not compose InvertedIndex and Normalizer objects

File: src/FileManager.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+    //if there was no specific tokenizer then we use the default splitTokenizer
+    //if there was no specific normalizer we use the lowercase normalizer as a default normalizer
+    public FileManager() {
+        this.normalizer = new UpperCaseNormalizer();
+        this.invertedIndex = new InvertedIndex(new SplitTokenizer(), normalizer);
+        setDefaultFilters();
+    }
```
Comment (MSPoulaei) [inline src/FileManager.java]: FileManager should not set defaults for language processing operations

File: src/FileManager.java
Comment (MSPoulaei) [inline src/FileManager.java:1]: rename the class name

File: src/dataStructures/ListCategory.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    private final Set<String> essentialFiles;
+    private final Set<String> optionalFiles;
+    private final Set<String> forbiddenFiles;
+    private boolean isFirstTime;
+
+    public ListCategory() {
+        essentialFiles = new HashSet<>();
+        optionalFiles = new HashSet<>();
+        forbiddenFiles = new HashSet<>();
+        isFirstTime = true;
```
Comment (MSPoulaei) [inline src/dataStructures/ListCategory.java:20]: move this boolean to where addToEssentialFile method was used

File: src/filter/normalizer/Normalizer.java
Comment (MSPoulaei) [inline src/filter/normalizer/Normalizer.java:1]: i think it's better we have a single String as input and normalize just one string at a method call
in your LoweCaseNormalizer and UpperCaseNormalizer there is so much duplication in for loop
we can avoid it by processing only one string.

File: src/searchMode/NormalSearch.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        super(database, query);
+    }
+
+    @Override
+    public Set<String> geAllDocuments() {
```
Comment (MSPoulaei) [inline src/searchMode/NormalSearch.java]: typo in method name

**Review by MSPoulaei — COMMENTED** [review]
required changes were commented in place.

File: src/searchMode/advancedSearch/AdvancedSearch.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import searchMode.Search;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class AdvancedSearch extends Search {
+    private final Set<String> queryWords;
+    private final ListCategory listCategory;
+
+
```
Comment (Ali-Athary) [inline src/searchMode/advancedSearch/AdvancedSearch.java]: نکته خط اضافه : اینجا دوتا خط اضافه گذاشتین. باید یدونه باشه.

File: src/dataStructures/ListCategory.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public void addToEssentialFile(Set<String> files) {
+        if (isFirstTime) {
+            essentialFiles.addAll(files);
+            isFirstTime = false;
+        }
+        else essentialFiles.retainAll(files);
+    }
```
Comment (Ali-Athary) [inline src/dataStructures/ListCategory.java:31]: اگه هیچ کلمه essential در وردی نباشد، این ست خالی میماند و جواب درست بر نمیگرداند.

File: src/searchMode/advancedSearch/AdvancedSearch.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            case ESSENTIAL:
+                listCategory.addToEssentialFile(files);
+                break;
+            case FORBIDDEN:
+                listCategory.addToForbiddenFile(files);
+                break;
+            case OPTIONAL:
+                listCategory.addToOptionalFile(files);
+                break;
+        }
```
Comment (Ali-Athary) [inline src/searchMode/advancedSearch/AdvancedSearch.java:45]: کلاس listCategory بهتره فقط یه تابع برای اضافه کردن داشته باشد و ListType رو به صورت ورودی بگیرد و switch case داخل خود کلاس باشد.

File: src/searchMode/NormalSearch.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
+package searchMode;
+
+import dataStructures.InvertedIndex;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class NormalSearch extends Search {
+    private Set<String> queryWords;
```
Comment (Ali-Athary) [inline src/searchMode/NormalSearch.java:11]: بعد تعریف کلاس یه خط خالی باشه.

File: src/reader/DocReader.java
Comment (Ali-Athary) [inline src/reader/DocReader.java:1]: شما برای این قسمت فرض کردین همه فایل ها txt هستند بهتره چک شود و توی نام کلاس ذکر شود.

File: src/dataStructures/InvertedIndex.java
Comment (Ali-Athary) [inline src/dataStructures/InvertedIndex.java:1]: بهتر بود با inverted index مثل یک ساختمان داده برخورد میکردین و فقط کار های مربوط به ذخیره و بازیابی اطلاعات رو داخلش انجام میدادید و کار های مربوط به Normalizer و Stemmer و ... رو بیرون کلاس انجام میداد.

**Review by Ali-Athary — CHANGES_REQUESTED** [review]
درست کنید.

**Review by Ali-Athary — APPROVED** [review]

**Review by MSPoulaei — APPROVED** [review]
all changes were applied!

File: src/DataManager.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import dataStructures.InvertedIndex;
+import filter.WordValidator;
+import filter.normalizer.Normalizer;
+import filter.normalizer.UpperCaseNormalizer;
+import filter.tokenizer.SplitTokenizer;
+import filter.tokenizer.Tokenizer;
+import reader.Reader;
+
```
Comment (Soltaniant) [inline src/DataManager.java:8]: remove the blank line.

File: src/DataManager.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.Map;
+
+public class DataManager {
+    private final InvertedIndex invertedIndex;
+    private final Normalizer normalizer;
+
+    //if there was no specific tokenizer then we use the default splitTokenizer
+    //if there was no specific normalizer we use the lowercase normalizer as a default normalizer
```
Comment (Soltaniant) [inline src/DataManager.java:16]: redundant comments.

File: src/DataManager.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+        invertedIndex.setDoStem(true);
+        invertedIndex.setWordValidator(new WordValidator());
+    }
+
+    public void createDatabase(Reader reader) {
+        Map<String, StringBuilder> file_texts = reader.getMapDocuments();
+        fillInvertedIndex(file_texts);
+    }
+
+    private void fillInvertedIndex(Map<String, StringBuilder> fileText) {
```
Comment (Soltaniant) [inline src/DataManager.java:63]: not a good name for parameter `fileText`

File: src/Main.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
+import filter.WordValidator;
+import filter.normalizer.LoweCaseNormalizer;
+import reader.TXTReader;
+import searchMode.Search;
+import filter.tokenizer.SplitTokenizer;
+import searchMode.advancedSearch.AdvancedSearch;
+
```
Comment (Soltaniant) [inline src/Main.java:7]: remove blank line

File: src/Main.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+import java.io.*;
+import java.util.*;
+
+public class Main {
+
+    public static final String path = "./textFiles";
+
+    public static void main(String[] args) throws IOException {
+        DataManager fileManager = new DataManager
```
Comment (Soltaniant) [inline src/Main.java:16]: DataManager fileManager!
we might need to rename these classes. are they specific enough?

File: src/Main.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+                (new SplitTokenizer(), new LoweCaseNormalizer());
```
Comment (Soltaniant) [inline src/Main.java:17]: Typo: LowCaseNormalizer

File: src/Main.java [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+        fileManager.setValidator(new WordValidator(true));
+        fileManager.setDoStem(true);
+        fileManager.createDatabase(new TXTReader(path));
+        String query = getQuery();
+        Search search = new AdvancedSearch(fileManager.getInvertedIndex(), query);
```
Comment (Soltaniant) [inline src/Main.java:22]: `fileManager.getInvertedIndex()`
it seems that we are getting InvertedIndex from a fileManager (actually a dataManager). it sound like a code smell to notice you might need to change something. maybe the architecture or simply a name.

File: src/Main.java [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+        search.printDocuments(search.getAllDocuments());
```
Comment (Soltaniant) [inline src/Main.java:23]: printing is not the responsibility of Search class. rename it or do some refactoring on your architecture.

File: src/DataManager.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+import filter.WordValidator;
+import filter.normalizer.Normalizer;
+import filter.normalizer.UpperCaseNormalizer;
+import filter.tokenizer.SplitTokenizer;
+import filter.tokenizer.Tokenizer;
+import reader.Reader;
+
+import java.util.Map;
+
+public class DataManager {
```
Comment (Soltaniant) [inline src/DataManager.java:11]: Read about Middle Man code smell.
`DataManager` has been a middle man.

File: src/DataManager.java [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+    public void setDoStem(boolean doStem) {
+        invertedIndex.setDoStem(doStem);
+    }
+
+    public void setDefaultFilters() {
+        invertedIndex.setDoStem(true);
+        invertedIndex.setWordValidator(new WordValidator());
+    }
+
```
Comment (Soltaniant) [inline src/DataManager.java:57]: middle man (obviously)

File: src/DataManager.java [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+    public void createDatabase(Reader reader) {
+        Map<String, StringBuilder> file_texts = reader.getMapDocuments();
+        fillInvertedIndex(file_texts);
+    }
+
+    private void fillInvertedIndex(Map<String, StringBuilder> fileText) {
+        for (String title : fileText.keySet()) {
+            invertedIndex.addFile(title, fileText.get(title));
```
Comment (Soltaniant) [inline src/DataManager.java:65]: `File` is not an appropriate name
+ you can rewrite code without `{}` brackets

File: src/Main.java [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+public class Main {
+
+    public static final String path = "./textFiles";
+
+    public static void main(String[] args) throws IOException {
+        DataManager fileManager = new DataManager
+                (new SplitTokenizer(), new LoweCaseNormalizer());
+        fileManager.setValidator(new WordValidator(true));
+        fileManager.setDoStem(true);
+        fileManager.createDatabase(new TXTReader(path));
```
Comment (Soltaniant) [inline src/Main.java:20]: **middleMan**

File: src/dataStructures/InvertedIndex.java [hunk 28]
```diff
@@ -0,0 +1,74 @@
+package dataStructures;
+
+import filter.WordValidator;
+import filter.normalizer.Normalizer;
+import filter.stemmer.Stemmer;
+import filter.tokenizer.Tokenizer;
+
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:7]: blank line

File: src/dataStructures/InvertedIndex.java [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class InvertedIndex {
+    private final Map<String, Set<String>> engine;
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:14]: `engine` might not be a good choice!

File: src/dataStructures/InvertedIndex.java [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+        this.wordValidator = validator;
+    }
+
+    public WordValidator getWordValidator() {
+        return wordValidator;
+    }
+
+    public Normalizer getNormalizer() {
+        return normalizer;
+    }
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:40]: Getters and setter are usually written last. (in the end lines of class)

File: src/dataStructures/InvertedIndex.java [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+
+    private void addWordToEngine(String root, String fileName) {
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:42]: `root` is not a good name.
also `fileName`

File: src/dataStructures/InvertedIndex.java [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+        if (!engine.containsKey(root)) {
+            engine.put(root, new HashSet<>());
+        }
+        engine.get(root).add(fileName);
+    }
+
+    public String getWordRoot(String word) {
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:49]: each time you use the `getWordRoot` method, you are creating a new instance of Stemmer()

File: src/dataStructures/InvertedIndex.java [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        return finalSetOfWords;
+    }
+
+
+    public void addFile(String title, StringBuilder fileText) {
+        for (String word : manipulateFile(fileText.toString())) {
+            addWordToEngine(word, title);
+        }
+    }
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:69]: the word `FILE` is not a good choice for describing our purpose.

File: src/dataStructures/InvertedIndex.java [hunk 34]
```diff
@@ -0,0 +1,74 @@
...
+        Set<String> tokenizedWords = new HashSet<>(tokenizer.tokenize(fileText));
+        Set<String> finalSetOfWords = new HashSet<>();
+        for (String word : tokenizedWords) {
+            word = normalizer.normalize(word);
+            if (wordValidator.isAcceptable(word))
+                finalSetOfWords.add(getWordRoot(word));
+        }
+        return finalSetOfWords;
+    }
+
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:63]: redundant blank line.
(exactly 1 line break is needed between methods)

File: src/dataStructures/InvertedIndex.java [hunk 35]
```diff
@@ -0,0 +1,74 @@
...
+
+    public void addFile(String title, StringBuilder fileText) {
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:65]: names!
`addFile`, `title`, `fileText`

File: src/dataStructures/InvertedIndex.java [hunk 36]
```diff
@@ -0,0 +1,74 @@
...
+            engine.put(root, new HashSet<>());
+        }
+        engine.get(root).add(fileName);
+    }
+
+    public String getWordRoot(String word) {
+        return doStem ? new Stemmer().getWordRoot(word) : word;
+    }
+
+    private Set<String> manipulateFile(String fileText) {
```
Comment (Soltaniant) [inline src/dataStructures/InvertedIndex.java:53]: `manipulateFile` -> not a good name!
also `fileText`

File: src/dataStructures/ListCategory.java [hunk 37]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class ListCategory {
+
+    private final Set<String> allFiles;
+    private final Set<String> essentialFiles;
+    private final Set<String> optionalFiles;
+    private final Set<String> forbiddenFiles;
```
Comment (Soltaniant) [inline src/dataStructures/ListCategory.java:11]: again: the word `File`
your invertedIndex should not be coupled to `File` in anyway!!

File: src/dataStructures/ListCategory.java [hunk 38]
```diff
@@ -0,0 +1,52 @@
+package dataStructures;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class ListCategory {
```
Comment (Soltaniant) [inline src/dataStructures/ListCategory.java:6]: not a good name

File: src/dataStructures/ListCategory.java [hunk 39]
```diff
@@ -0,0 +1,52 @@
...
+    public void addToEssentialFile(Set<String> files) {
+        hasEssentialWords = true;
+        if (isFirstTime) {
+            essentialFiles.addAll(files);
+            isFirstTime = false;
+        }
+        else essentialFiles.retainAll(files);
+    }
+
+    public void addToOptionalFile(Set<String> files) {
```
Comment (Soltaniant) [inline src/dataStructures/ListCategory.java:33]: addToOptionalFile**s**
addToForbiddenFile**s**

File: src/filter/WordValidator.java [hunk 40]
```diff
@@ -0,0 +1,29 @@
+package filter;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class WordValidator {
```
Comment (Soltaniant) [inline src/filter/WordValidator.java:7]: is `WordValidator` name, good enough?

File: src/reader/Reader.java [hunk 41]
```diff
@@ -0,0 +1,20 @@
+package reader;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class Reader {
```
Comment (Soltaniant) [inline src/reader/Reader.java:6]: `Reader` name does not essentially imply fileReader!

File: src/searchMode/advancedSearch/ListType.java [hunk 42]
```diff
@@ -0,0 +1,7 @@
+package searchMode.advancedSearch;
+
+public enum ListType {
```
Comment (Soltaniant) [inline src/searchMode/advancedSearch/ListType.java:3]: not a good name

**Review by Soltaniant — COMMENTED** [review]
دمتون گرم و خسته نباشید :+1: 
در مورد نام‌گذاری‌ها یه مقدار مشکل داشتیم. نام‌گذاریمون باید گویا تر باشه و به بهترین حالت بتونه عملکرد اون متود، کلاس یا متغیر رو بیان کنه.

**Pull request merged**